### PR TITLE
Fix/issue 2709 - Fix miscalculation tax from TaxJar and decided to use nexus address

### LINKED ISSUE #2709: Incorrect tax rates for CO located merchants: city tax not calculated
I recently had a merchant reach out to let us know that taxes are not being calculated as expected for their store in Colorado. 

The merchant's store is located in: `80907` and the city tax is not added unless customer is is also located in `80907`. 

**Steps to replicate:**

1. Head to TaxJar Sales Tax Calculator to check the sales tax for Aurora, CO: https://www.taxjar.com/sales-tax-calculator
2. Find TaxJar shows a tax rate of 8.000% 
3. Enable automated taxes using WooCommerce Shipping & Tax and set shop location to Colorado Springs, CO (`80907`)
4. Add a product to you cart and head to the checkout page - enter an Aurora, CO address at checkout with zip code `80013`
5. Find tax rate seen at checkout is only `4.25%` (.25% county + 2.9% state + 1.1% special tax rate)

I ran the zip 80013 through the TaxJar sales tax calculator and I found the rate of 8.0%: 

Screen Shot 2023-12-22 at 3 15 05 PM

 
TaxJar Sales Tax Calculator: https://www.taxjar.com/sales-tax-calculator

 
However, when I run the request using https://developers.taxjar.com/demo/ with the from zip code of 80907 to the zip code 80013, I see no city tax is calculated (the missing 3.75%): 
 
Screen Shot 2023-12-22 at 3 17 31 PM

I believe this issue is with TaxJar, a 3rd party vendor, but I wanted to flag this here.

### PR DESCRIPTION
**Pull request by iyut**
## Description
Based on the discussion with TaxJar, it's rather to omit the `from_*` from the API request and use `nexus_addresses` instead. And after doing a couple of test, this method seems fixing the tax issue and I dont see any other issue when implementing it.

### Related issue(s)

Fixes #2709 

### Steps to reproduce & screenshots/GIFs

1. Head to TaxJar Sales Tax Calculator to check the sales tax for Aurora, CO: https://www.taxjar.com/sales-tax-calculator
2. Find TaxJar shows a tax rate of 8.000% 
3. Enable automated taxes using WooCommerce Shipping & Tax and set shop location to Colorado Springs, CO (`80907`)
4. Add a product to you cart and head to the checkout page - enter an Aurora, CO address at checkout with zip code `80013`
5. Find tax rate seen at checkout is `8%`
Screenshot 2024-02-07 at 09 39 46

### Checklist

- [ ] unit tests

- [x] `changelog.txt` entry added
- [x] `readme.txt` entry added

### CODE REVIEW

**Review by kloon — COMMENTED** [review]
Just saw this and thought I'd ask.

https://github.com/Automattic/woocommerce-services/pull/2716/files#diff-beb8b82c50d2f6d2459a47fb37de080481a77d95fca469d8b32c6140e25755faR1080-R1088

The existing code actually uses the to_* fields as the nexus address, we should be using the from_ fields as the nexus address from my understanding, IE the store address.

Also if we switch to the from_* fields, we should unset the fields afterwards and not before as the current code does.

**Comment by iyut** [comment]
@kloon I have tried to dig more info about this. And found [this article](https://www.taxjar.com/sales-tax/origin-based-and-destination-based-sales-tax). It seems there is more to it. The way we charge the tax might be different on each state. Some of them might use origin-based tax and the other are destination-based tax.

But based on [this article](https://www.bench.co/blog/tax-tips/sales-tax-nexus), it's said that : "a majority of states are 'destination-based,' meaning they require merchants to collect sales tax based on the buyer’s ship-to location."

I have also tried to change the nexus addresses info to use `from_*` ( store address ). The result is when shipping :

- from CO to CO : It will return the tax 8%.
- from CO to CA : It doesnt return any tax.

But if I use `to_*` ( shipping address ). The result is : 

- from CO to CO : It will return the tax 8%.
- from CO to CA : It will return tax 9.5%.

Should we ask them more detail about `nexus_addresses` implementation?

Here is the request that was generated just in case they need it : 

Using `from_*` :
`{"to_country":"US","to_state":"CA","to_zip":"90210","to_city":"Beverly Hills","to_street":false,"shipping":"10","plugin":"woo","nexus_addresses":[{"street":"1275 Fraser St","city":"Aurora","state":"CO","country":"US","zip":"80011"}],"line_items":[{"id":"14-aab3238922bcc25a6f606eb525ffdc56","quantity":1,"product_tax_code":"","unit_price":"18","discount":"0"},{"id":"31-c16a5320fa475530d9583c34fd356ef5","quantity":1,"product_tax_code":"","unit_price":"18","discount":"0"}]} `

using `to_*` : 
`{"to_country":"US","to_state":"CA","to_zip":"90210","to_city":"Beverly Hills","to_street":false,"shipping":"10","plugin":"woo","nexus_addresses":[{"street":false,"city":"Beverly Hills","state":"CA","country":"US","zip":"90210"}],"line_items":[{"id":"14-aab3238922bcc25a6f606eb525ffdc56","quantity":1,"product_tax_code":"","unit_price":"18","discount":"0"},{"id":"31-c16a5320fa475530d9583c34fd356ef5","quantity":1,"product_tax_code":"","unit_price":"18","discount":"0"}]}`

**Comment by kloon** [comment]
Thanks for the feedback @iyut, I have relayed the question to our tax partner for clarification.

**Comment by kloon** [comment]
@iyut I have received feedback from TaxJar:

> 1. Sales Tax Calculation with FROM Address: If you don't have the FROM address when calculating sales tax, you can obtain tax rates for a given location by using the /v2/rates endpoint and providing the customer's address. However, this method may provide only a rough estimate, especially if the seller doesn't have nexus in that state. For the most accurate calculation, it's recommended to use the FROM address of the merchant selling the product. This can be done by utilizing the from_ or nexus_addresses parameters with the /v2/taxes endpoint. This approach ensures the most accurate sales tax calculation and determines if sales tax is due.
> 2. Differentiating Between Origin-Based and Destination-Based Tax States: Our API automatically identifies whether a state is origin-based or destination-based. You can also verify this information on the Dashboard, where it is displayed in the upper left corner when reviewing any state [0].
Regarding your test from CO to CA that didn't return a calculation, I'll investigate further to understand why this occurred. Did you encounter any specific error message during the test?

**Comment by iyut** [comment]
Hi @kloon ,

I have done more testing into TaxJar API, and it seems even if we're using nexus addresses as a from. It will still show a wrong tax rate from Colorado Springs, CO, 80907 to Aurora, CO, 80012. It's still return 4.25% tax rate. 

`Requesting: taxjar/v2/taxes - {"to_country":"US","to_state":"CO","to_zip":"80012","to_city":"Aurora","to_street":"800 S Abilene St, Aurora, CO 80012, United States","shipping":"10","plugin":"woo","nexus_addresses":[{"street":"2300 N Wahsatch Ave","city":"Colorado Springs","state":"CO","country":"US","zip":"80907"}],"line_items":[{"id":"31-c16a5320fa475530d9583c34fd356ef5","quantity":1,"product_tax_code":"","unit_price":"18","discount":"0"}]}`

`Received: {"tax":{"amount_to_collect":0.77,"breakdown":{"city_tax_collectable":0.0,"city_tax_rate":0.0,"city_taxable_amount":0.0,"combined_tax_rate":0.0425,"county_tax_collectable":0.05,"county_tax_rate":0.0025,"county_taxable_amount":18.0,"line_items":[{"city_amount":0.0,"city_tax_rate":0.0,"city_taxable_amount":0.0,"combined_tax_rate":0.0425,"county_amount":0.05,"county_tax_rate":0.0025,"county_taxable_amount":18.0,"id":"31-c16a5320fa475530d9583c34fd356ef5","special_district_amount":0.2,"special_district_taxable_amount":18.0,"special_tax_rate":0.011,"state_amount":0.52,"state_sales_tax_rate":0.029,"state_taxable_amount":18.0,"tax_collectable":0.77,"taxable_amount":18.0}],"special_district_tax_collectable":0.2,"special_district_taxable_amount":18.0,"special_tax_rate":0.011,"state_tax_collectable":0.52,"state_tax_rate":0.029,"state_taxable_amount":18.0,"tax_collectable":0.77,"taxable_amount":18.0},"freight_taxable":false,"has_nexus":true,"jurisdictions":{"city":"AURORA","country":"US","county":"ARAPAHOE","state":"CO"},"order_total_amount":28.0,"rate":0.0425,"shipping":10.0,"tax_source":"destination","taxable_amount":18.0}}`

I'm really wondering how they calculate it that way.

Regarding their question : _Did you encounter any specific error message during the test?_
There is no error specified when testing it. It was return 0.

But I double test the tax rates again today from CO to CA:
When not using nexus addresses, it returns the tax rate 9.5%.
But when using nexus addresses it doesn't return tax rates.

**Comment by kloon** [comment]
Thanks for checking again @iyut 

I have relayed this back to TaxJar support again, in the interim, perhaps we do a solution like we did for Canada, but for Colorado while we iron out this with TaxJar on using the Nexus Address.

**Comment by iyut** [comment]
Thanks for the suggestion @kloon , In fact, that's what our team have discussed yesterday. Applying nexus address to all store address seems risky and might have a potential for regression.

File: classes/class-wc-connect-taxjar-integration.php [hunk 1]
```diff
@@ -978,6 +978,62 @@ function( $line_item ) use ( $new_tax_rate ) {
...
 	}
 
+	/**
+	 * Implement nexus address for a certain state or country.
+	 *
+	 * @param array $body Request body.
+	 *
+	 * @return array
+	 */
+	public function use_nexus_address( $body ) {
```
Comment (dustinparker) [inline classes/class-wc-connect-taxjar-integration.php]:
```suggestion
	public function maybe_apply_taxjar_nexus_addresses_workaround( $body ) {
```

File: classes/class-wc-connect-taxjar-integration.php [hunk 2]
```diff
@@ -1063,34 +1119,8 @@ public function calculate_tax( $options = array() ) {
...
-				array(
-					'street'  => $body['to_street'],
-					'city'    => $body['to_city'],
-					'state'   => $body['to_state'],
-					'country' => $body['to_country'],
-					'zip'     => $body['to_zip'],
-				),
-			);
-		}
+		$body = $this->use_nexus_address( $body );
```
Comment (dustinparker) [inline classes/class-wc-connect-taxjar-integration.php]:
```suggestion
		$body = $this->maybe_apply_taxjar_nexus_addresses_workaround( $body );
```

File: classes/class-wc-connect-taxjar-integration.php [hunk 3]
```diff
@@ -978,6 +978,62 @@ function( $line_item ) use ( $new_tax_rate ) {
...
+				'from_street',
+			);
+
+			foreach ( $params_to_unset as $param ) {
+				unset( $body[ $param ] );
+			}
+		}
+
+		return $body;
+	}
```
Comment (dustinparker) [inline classes/class-wc-connect-taxjar-integration.php]:
```suggestion
	/**
	 * Maybe apply a temporary workaround for the TaxJar API to get the correct rates for
	 * specific edge cases.
	 *
	 * For these specific edge cases a "nexus_addresses" element needs to be added to the
	 * TaxJar request body and the "from" address needs to be removed from it in order to
	 * get the correct rates. This is due to a limitation/miscalculation at the TaxJar API.
	 *
	 * This method adds the "nexus_addresses" element to the request body and unsets the "from"
	 * address elements if the workaround is enabled and an address case is matched.
	 *
	 * New edge cases can be added to the $cases array as needed.
	 *
	 * @param array $body Request body.
	 *
	 * @return array
	 */
	public function maybe_apply_taxjar_nexus_addresses_workaround( $body ) {
		if ( true !== apply_filters( 'woocommerce_apply_taxjar_nexus_addresses_workaround', true ) ) {
			return $body;
		}

		$cases = array(
			'CA-QC' => array(
				'to_country'   => 'CA',
				'to_state'     => 'QC',
				'from_country' => 'CA',
			),
			'US-CO' => array(
				'to_country'   => 'US',
				'to_state'     => 'CO',
				'from_country' => 'US',
			),
		);

		foreach ( $cases as $case ) {
			if ( $case['to_country'] !== $body['to_country'] ||
				$case['to_state'] !== $body['to_state'] ||
				$case['from_country'] !== $body['from_country'] ) {
				continue;
			}

			$body['nexus_addresses'] = array(
				array(
					'street'  => $body['to_street'],
					'city'    => $body['to_city'],
					'state'   => $body['to_state'],
					'country' => $body['to_country'],
					'zip'     => $body['to_zip'],
				),
			);

			$params_to_unset = array(
				'from_country',
				'from_state',
				'from_zip',
				'from_city',
				'from_street',
			);

			foreach ( $params_to_unset as $param ) {
				unset( $body[ $param ] );
			}

			break;
		}

		return $body;
	}
```

**Review by dustinparker — CHANGES_REQUESTED** [review]
This fixes the issue for Colorado taxes. Nice work! We just need to bump the WC versions in the main plugin file and readme file. I also suggested a few changes to help future devs better understand what we are doing with this change.

File: classes/class-wc-connect-taxjar-integration.php (hunk 3 above)
Comment (iyut) [reply classes/class-wc-connect-taxjar-integration.php]: Thanks for the suggestion @dustinparker . I find this code more efficient!

**Review by dustinparker — APPROVED** [review]
Looks good to go!

**Pull request merged**